modules/videoio/src/cap_msmf.cpp

```diff
@@ -997,71 +1236,181 @@ bool CvCapture_MSMF::openFinalize_(const VideoCaptureParameters* params)
+        if (value != CV_8S && value != CV_16S && value != CV_32S && value != CV_32F)
         {
-            CV_LOG_WARNING(NULL, "videoio(MSMF): EOS signal. Capture stream is lost");
+            CV_LOG_ERROR(NULL, "VIDEOIO/MSMF: CAP_PROP_AUDIO_DATA_DEPTH parameter value is invalid/unsupported: " << value);
             return false;
+        } else
+        {
+            outputAudioFormat = value;
         }
-        sampleTime = reader->m_lastSampleTimestamp;
-        return true;
     }
-    else if (isOpen)
+    if (params.has(CAP_PROP_SYNC_LAST_FRAME))
     {
-        DWORD streamIndex, flags;
-        videoSample.Release();
-        HRESULT hr;
-        for(;;)
+        int value = static_cast<int>(params.get<double>(CAP_PROP_SYNC_LAST_FRAME));
+        if (value != 0 && value != 1)
+        {
+            CV_LOG_ERROR(NULL, "VIDEOIO/MSMF: CAP_PROP_SYNC_LAST_FRAME parameter value is invalid/unsupported: " << value);
+            return false;
+        } else
+        {
+            syncLastFrame = (value == 1) ? true : false;
+        }
+    }
+    return true;
+}
+
+bool CvCapture_MSMF::grabVideoFrame()
+{
+    DWORD streamIndex,  flags;
+    HRESULT hr;
+    videoSample.Release();
+    for (;;)
+    {
+        CV_TRACE_REGION("ReadSample");
+        if (!SUCCEEDED(hr = videoFileSource->ReadSample(
+            dwVideoStreamIndex, // Stream index.
+            0,             // Flags.
+            &streamIndex,  // Receives the actual stream index.
+            &flags,        // Receives status flags.
+            &sampleTime,   // Receives the time stamp.
+            &videoSample   // Receives the sample or NULL.
+        )))
+            break;
+        if (streamIndex != dwVideoStreamIndex)
+            break;
+        if (flags & (MF_SOURCE_READERF_ERROR | MF_SOURCE_READERF_ALLEFFECTSREMOVED | MF_SOURCE_READERF_ENDOFSTREAM))
+            break;
+        if (videoSample)
+            break;
+        if (flags & MF_SOURCE_READERF_STREAMTICK)
+        {
+            CV_LOG_DEBUG(NULL, "videoio(MSMF): Stream tick detected. Retrying to grab the frame");
+        }
+    }
+    if (SUCCEEDED(hr))
+    {
+        if (streamIndex != dwVideoStreamIndex)
+        {
+            CV_LOG_DEBUG(NULL, "videoio(MSMF): Wrong stream read. Abort capturing");
+            close();
+        }
+        else if (flags & MF_SOURCE_READERF_ERROR)
+        {
+            CV_LOG_DEBUG(NULL, "videoio(MSMF): Stream reading error. Abort capturing");
+            close();
+        }
+        else if (flags & MF_SOURCE_READERF_ALLEFFECTSREMOVED)
+        {
+            CV_LOG_DEBUG(NULL, "videoio(MSMF): Stream decoding error. Abort capturing");
+            close();
+        }
+        else if (flags & MF_SOURCE_READERF_ENDOFSTREAM)
+        {
+            sampleTime += frameStep;
+            CV_LOG_DEBUG(NULL, "videoio(MSMF): End of stream detected");
+        }
+        else
+        {
+            sampleTime += frameStep;
```

Comment:
> `sampleTime += frameStep;`

This may provide inaccurate values.
Use timestamp value from ReadSample. See code related to `m_lastSampleTimestamp`.

Reply:
Timestamp value from Read Sample does not always indicate the end of the captured sample. In the case of {"mov", "H 264", 30. f, CAP_MSMF}The time stamp in ReadSample is set at the beginning of the captured frame and for correct operation it is necessary to calculate SampleTime= + frameStep

Reply:
> > `sampleTime += frameStep;`
> 
> This may provide inaccurate values.
> Use timestamp value from ReadSample. See code related to `m_lastSampleTimestamp`.

Timestamp value from Read Sample does not always indicate the end of the captured sample. In the case of {"mov", "H 264", 30. f, CAP_MSMF}The time stamp in ReadSample is set at the beginning of the captured frame and for correct operation it is necessary to calculate SampleTime= + frameStep

Reply:
> the end of the captured sample
> end

This is not needed at all and can NOT be properly determined for VFR videos (without reading of the next frame).

Use provided value from the decoder/demuxer.
Compare results with FFmpeg backend.